File: src/factories/internal/metadata/emplace_metadata_tuple.ts
import ts from "typescript";

import { Metadata } from "../../../schemas/metadata/Metadata";
import { MetadataTupleType } from "../../../schemas/metadata/MetadataTupleType";

import { Writable } from "../../../typings/Writable";

import { ArrayUtil } from "../../../utils/ArrayUtil";

import { IMetadataIteratorProps } from "./IMetadataIteratorProps";
import { explore_metadata } from "./explore_metadata";

export const emplace_metadata_tuple = (
  props: IMetadataIteratorProps<ts.TupleType>,
): MetadataTupleType => {
  // CHECK EXISTENCE
  const [tuple, newbie, closure] = props.collection.emplaceTuple(
    props.checker,
    props.type,
  );
  ArrayUtil.add(tuple.nullables, props.metadata.nullable);
  if (newbie === false) return tuple;

  // CONSTRUCT ELEMENT TYPES
  const flagList: readonly ts.ElementFlags[] =
    props.type.elementFlags ??
    (props.type.target as ts.TupleType)?.elementFlags ??
    [];
  const elements: Metadata[] = props.checker
    .getTypeArguments(props.type as ts.TypeReference)
    .map((elem, i) => {
      const child: Metadata = explore_metadata({
        ...props,
        type: elem,
        explore: {
          ...props.explore,
          nested: tuple,
          aliased: false,
          escaped: false,
        },
        intersected: false,
      });

      // CHECK OPTIONAL
      const flag: ts.ElementFlags | undefined = flagList[i];
      if (flag === ts.ElementFlags.Optional) Writable(child).optional = true;

      // REST TYPE
      if (flag !== ts.ElementFlags.Rest) return child;
      const wrapper: Metadata = Metadata.initialize();
      Writable(wrapper).rest = child;
      return wrapper;
    });
  closure(elements);

  return tuple;
};

File: packages/typia/native/core/factories/internal/metadata/emplace_metadata_tuple.go
package metadata

import (
	nativechecker "github.com/microsoft/typescript-go/shim/checker"
	schemametadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
)

func Emplace_metadata_tuple(props IMetadataIteratorProps) *schemametadata.MetadataTupleType {
	tuple, newbie, closure := props.Components.EmplaceTuple(
		props.Checker,
		props.Type,
	)
	metadata_array_util_add_bool(&tuple.Nullables, props.Metadata.Nullable)
	if newbie == false {
		return tuple
	}

	flagList := []nativechecker.ElementFlags{}
	if props.Type != nil && nativechecker.IsTupleType(props.Type) {
		flagList = props.Type.TargetTupleType().ElementFlags()
	}
	args := []*nativechecker.Type{}
	if props.Checker != nil {
		args = metadata_get_type_arguments(props.Checker, props.Type)
	}
	elements := make([]*schemametadata.MetadataSchema, 0, len(args))
	for i, elem := range args {
		explore := props.Explore
		explore.Nested = tuple
		explore.Aliased = false
		explore.Escaped = false
		child := Explore_metadata(Explore_metadata_IProps{
			Options:     props.Options,
			Checker:     props.Checker,
			Components:  props.Components,
			Errors:      props.Errors,
			Type:        elem,
			Explore:     explore,
			Intersected: false,
		})
		var flag nativechecker.ElementFlags
		if i < len(flagList) {
			flag = flagList[i]
		}
		if flag == nativechecker.ElementFlagsOptional {
			child.Optional = true
		}
		if flag != nativechecker.ElementFlagsRest {
			elements = append(elements, child)
			continue
		}
		wrapper := schemametadata.MetadataSchema_initialize()
		wrapper.Rest = child
		elements = append(elements, wrapper)
	}
	closure(elements)
	return tuple
}
